Big Σs and Big Πs over finite integer ranges

hacklex opened the pull request:
Added the general `fold` operation for CommMonoids together with a lemma that establish the equality of such fold to a `foldm_snoc` of the corresponding `seq`.

This will be useful when I further introduce FStar.Seq.Matrix and establish a link between double `fold` and a `foldm_snoc` over a `matrix_seq`.

On its own, this module might be useful whenever one does need a finite sum, but does not want to construct an explicit seq for it.

nikswamy commented on ulib/FStar.Algebra.CommMonoid.Fold.fst:
```diff
@@ -0,0 +1,129 @@
+   a sum or a product, depending on the monoid operation. *)
+ 
+(* We refine multiplication a bit to make proofs smoothier *)
+let ( *) (x y: int) : (t:int{(x>0 /\ y>0) ==> t>0}) = op_Multiply x y
```
I'm not sure this re-definition of multiplication is worth it. Is it really making proofs simpler? I would just open `FStar.Mul` and use the `*` that it defines.

nikswamy commented on ulib/FStar.Algebra.CommMonoid.Fold.fst:
```diff
@@ -0,0 +1,129 @@
+type inbetween (x: int) (y: not_less_than x) = (t: int{t>=x && t<=y})
+type counter_of_range (x: int) (y: not_less_than x) = (t: nat{t<(y+1-x)})
+
+let range_count (x: int) (y: not_less_than x) : pos = (y+1)-x
```
These definitions are similar to the ones in a previous PR ... can we put them somewhere common and re-use them?

For example, can we add a module like FStar.Intervals or something like that?

As a side-remark: What's a bit surprising about `inbetween` is that we almost always work with half-open intervals, but this one defines a closed interval.

hacklex replied:
yep, we certainly should. Perhaps add both closed and half-closed versions too. I already noticed this appears occasionally in several ulib files. Perhaps everyone eventually comes up with his own range type :)

nikswamy commented on ulib/FStar.Algebra.CommMonoid.Fold.fst:
```diff
@@ -0,0 +1,129 @@
+             (nk: not_less_than n0) 
+             (expr: (inbetween n0 nk) -> c) 
+  : Pure c (requires True)  
+           (ensures (fun (x:c) -> ((nk = n0) ==> (x == expr nk)))) 
```
Surprising to include the base case of the recursion in the ensures clause. I would just write this type as `Tot c (decreases nk-n0)`

nikswamy commented on ulib/FStar.Algebra.CommMonoid.Fold.fst:
```diff
@@ -0,0 +1,129 @@
+                             (expr: (inbetween n0 nk) -> c)
+  : Lemma (ensures fold cm n0 nk expr `eq.eq` 
+                   foldm_snoc cm (init (range_count n0 nk) 
+                                       (init_func_from_expr expr n0 nk)))
```
Nice!

hacklex commented on ulib/FStar.Algebra.CommMonoid.Fold.fst:
```diff
@@ -0,0 +1,129 @@
+  if nk = n0 then expr nk
+  else (fold cm n0 (nk-1) expr) `cm.mult` expr nk
+
+(* This lemma establishes the definitional equality of the fold 
```
Yes thank you. I messed up with terminology, thanks for letting me know! Will fix!

hacklex commented:
Alright. I've extracted intervals to a separate module, and reworked my new Fold module to use them.

I see you pointed out certain redundancies. I remember that I added that ensures clause for a reason (some proof went faster/easier with it), but currently (in the scope of current PR) this clause makes no difference. 

So perhaps we should first integrate my modules without refinements (open FStar.Mul instead of defining my own mul, and remove that weird ensures clause) -- and when I make the next pull request with a module that would benefit from these refinements, we discuss the matter again with concrete usage example in mind?

I'll prepare a commit shortly that fixes all the unpleasant things you mentioned.

hacklex commented:
Alright, done. 

Now we have FStar.IntegerIntervals that defines all kinds of integer intervals in full generality, along with some utils.

Perhaps we would want to move existing definitions to use this new module, but I'm not sure how to approach that. 

Anyway, my new `CommMonoid.Fold` module now relies to these generalized intervals definitions, and overall looks quite pretty. I think this one's ready to be included in `ulib` :)

hacklex commented:
Alright, done. 

Now we have FStar.IntegerIntervals that defines all kinds of integer intervals in full generality, along with some utils.

Perhaps we would want to move existing definitions to use this new module, but I'm not sure how to approach that. 

Anyway, my new `CommMonoid.Fold` module now relies to these generalized intervals definitions, and overall looks quite pretty. I think this one's ready to be included in `ulib` :)

nikswamy commented:
Thank you!
